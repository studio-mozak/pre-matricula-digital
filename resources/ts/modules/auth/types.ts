export interface Auth {
  level: number;
  name: string;
  schools: string[];
}
