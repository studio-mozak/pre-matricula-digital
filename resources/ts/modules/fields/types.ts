import { Nullable } from '@/types';

export type SetValues = (fields: Record<string, unknown>) => void;

export interface FieldPageOption {
  id?: string;
  name: Nullable<string>;
  weight: number;
}

export interface FieldPage {
  id: Nullable<string>;
  name: Nullable<string>;
  description: Nullable<string>;
  group: Nullable<'RESPONSIBLE' | 'STUDENT'>;
  type: Nullable<string>;
  internal: boolean;
  options: FieldPageOption[];
  deleteOptions?: string[];
  required?: boolean;
}

export interface ErrorResponse {
  message: string;
}
