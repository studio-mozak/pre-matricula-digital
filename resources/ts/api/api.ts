import axios, { AxiosInstance } from 'axios';

const token = window.config ? window.config.token : '';

export const http: AxiosInstance = axios.create({
  baseURL: String(import.meta.env.VITE_APP_API_URL || ''),
  headers: {
    Authorization: `Token ${token}`,
  },
});

axios.interceptors.response.use(
  (res) => {
    /**
     * If the response from the server has the "errors" object or the object "data" is empty,
     * here it is forced to go to the catch block automatically.
     */
    if (
      (res.data.errors || Object.keys(res.data).length === 0) &&
      !(res.data instanceof Blob)
    ) {
      return Promise.reject({
        response: {
          data: res.data,
        },
      });
    }

    return res;
  },
  (err) => Promise.reject(err)
);

export const rest = <T>(
  type: 'get' | 'post' | 'put' | 'delete',
  url: string,
  params?: object
): Promise<T> => http[type](url, params);

export const graphql = <T>(data: object, config?: object): Promise<T> =>
  axios.post('/graphql', data, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
    ...config,
  });
