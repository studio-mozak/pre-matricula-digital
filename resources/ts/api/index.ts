export * from './api';
export * as City from './services/graphql/city';
export * as Resource from './services/graphql/resource';
