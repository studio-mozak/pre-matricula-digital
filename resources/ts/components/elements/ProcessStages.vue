<template>
  <div class="row">
    <template v-for="stage in stages" :key="stage.id">
      <div class="col-12 col-md-6 mt-3 mt-sm-2">
        <process-stage-card :stage="stage" />
      </div>
    </template>
  </div>
</template>

<script setup lang="ts">
import { Processes, Stages } from '@/types';
import ProcessStageCard from '@/components/elements/ProcessStageCard.vue';
import { computed } from 'vue';

const props = withDefaults(
  defineProps<{
    process: Processes;
  }>(),
  {
    process: () => ({} as Processes),
  }
);

const stages = computed(() => {
  const { stages } = props.process;

  if (!stages) return [];

  const processesOpened: Stages[] = [];
  const processesNotOpened: Stages[] = [];
  stages.forEach((stage) => {
    if (stage.status === 'OPEN') {
      processesOpened.push(stage);
    } else {
      processesNotOpened.push(stage);
    }
  });
  processesOpened.sortBy('startAt').reverse();
  processesNotOpened.sortBy('startAt').reverse();
  return processesOpened.concat(processesNotOpened);
});
</script>
