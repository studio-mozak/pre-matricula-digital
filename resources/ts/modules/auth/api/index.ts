/**
 * Exporting the Axios Instance
 */
export { http, graphql } from '@/api';

/**
 * Exporting all modules
 */
export * as AuthRest from '@/modules/auth/api/services/rest/auth';
