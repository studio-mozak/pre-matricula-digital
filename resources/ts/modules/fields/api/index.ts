/**
 * Exporting the Axios Instance
 */
export { http, graphql } from '@/api';

/**
 * Exporting all modules
 */
export * as Field from './services/graphql/field';
