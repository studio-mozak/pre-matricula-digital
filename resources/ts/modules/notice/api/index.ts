/**
 * Exporting the Axios Instance
 */
export { http, graphql } from '@/api';

/**
 * Exporting all modules
 */
export * as Notice from './services/graphql/notice';
