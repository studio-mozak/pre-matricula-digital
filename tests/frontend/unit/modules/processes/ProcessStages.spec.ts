import { VueWrapper, flushPromises, mount } from '@vue/test-utils';
import { Process as ProcessApi } from '@/modules/processes/api';
import { ProcessStageResponse } from '@/modules/processes/types';
import ProcessStages from '@/modules/processes/components/ProcessStages.vue';
import { createTestingPinia } from '@pinia/testing';
import { expect, vi } from 'vitest';
import { analytics } from '@/packages';
import { AnalyticsInstance } from 'analytics';

interface ProcessStagesProps {
  id?: string;
  newProcess: string;
}

vi.mock('@/modules/processes/api');
vi.mock('vue-router/dist/vue-router.mjs', () => ({
  useRouter: () => ({
    push: spyPush,
  }),
}));
vi.mock('@/packages');

const spyPush = vi.fn();

const processMock: ProcessStageResponse = {
  process: {
    id: '1',
    name: 'Ed. Infantil (Nível III, Pré-Escola) Ens. Fundamental (1° ao 9° ano)',
  },
  stages: [
    {
      allowWaitingList: true,
      endAt: '2021-03-07 23:59:00',
      endHourAt: '2021-03-07 23:59:00',
      id: '1',
      observation: '',
      radius: 1500,
      renewalAtSameSchool: false,
      startAt: '2021-02-17 14:00:00',
      startHourAt: '2021-02-17 14:00:00',
      type: 'REGISTRATION',
      allowSearch: false,
      restrictionType: 'NONE',
    },
  ],
};

const onMounted = async (props: ProcessStagesProps) => {
  vi.mocked(ProcessApi.listStages).mockResolvedValueOnce(processMock);

  vi.mocked(analytics).mockImplementation(
    () =>
      ({
        page: vi.fn(),
      } as unknown as AnalyticsInstance)
  );

  const wrapper = mount(ProcessStages, {
    props,
    global: {
      plugins: [createTestingPinia()],
      stubs: ['x-field'],
    },
  });

  await wrapper.vm.$nextTick();

  await flushPromises();

  return wrapper;
};

describe('ProcessStages', () => {
  let wrapper: VueWrapper<InstanceType<typeof ProcessStages>>;

  beforeEach(async () => {
    wrapper = await onMounted({
      newProcess: 'false',
      id: '1',
    });

    await wrapper.vm.$nextTick();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test('the component exists', () => {
    expect(wrapper).toBeTruthy();
  });

  test('should match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  test('the initial variables must be loaded with the correct data', () => {
    expect(wrapper.vm.stages).toMatchObject(processMock.stages);
    expect(wrapper.vm.data).toMatchObject(processMock);
  });

  describe('addStage', () => {
    test('should call addStage method', async () => {
      wrapper.vm.addStage = vi.fn();

      wrapper.vm.addStage();

      await wrapper.vm.$nextTick();

      expect(wrapper.vm.addStage).toHaveBeenCalled();
      expect(wrapper.vm.addStage).toHaveBeenCalledTimes(1);
    });

    test('should call addStage method and push the stages array', async () => {
      expect(wrapper.vm.stages).toHaveLength(1);

      wrapper.vm.addStage();

      await wrapper.vm.$nextTick();

      expect(wrapper.vm.stages).toHaveLength(2);
    });
  });

  describe('removeStage', () => {
    test('should call removeStage method', async () => {
      wrapper.vm.removeStage = vi.fn();

      wrapper.vm.removeStage();

      await wrapper.vm.$nextTick();

      expect(wrapper.vm.removeStage).toHaveBeenCalled();
      expect(wrapper.vm.removeStage).toHaveBeenCalledTimes(1);
    });

    test('should call removeStage method and remove 1 item of the stages array', async () => {
      expect(wrapper.vm.stages).toHaveLength(2);

      wrapper.vm.removeStage();

      await wrapper.vm.$nextTick();

      expect(wrapper.vm.removeStage).toHaveLength(1);
    });
  });

  test('if I click on "go back" button, the user should be redirected to previous page', async () => {
    const button = wrapper.find('[data-test="button-back"]');

    await button.trigger('click');

    expect(spyPush).toBeCalled();
    expect(spyPush).toBeCalledTimes(1);
    expect(spyPush).toBeCalledWith({
      name: 'process.fields',
      params: {
        id: '1',
        newProcess: 'false',
      },
    });
  });

  describe('submit', () => {
    test('should call submit method successfully', async () => {
      vi.mocked(ProcessApi.postStages).mockResolvedValueOnce();

      wrapper.vm.submit();

      await wrapper.vm.$nextTick();

      await flushPromises();

      expect(spyPush).toHaveBeenCalledWith({
        name: 'process.check',
        params: {
          id: '1',
          newProcess: 'false',
        },
      });
    });

    test('should call submit method unsuccessfully', async () => {
      vi.mocked(ProcessApi.postStages).mockRejectedValueOnce();

      wrapper.vm.submit();

      await wrapper.vm.$nextTick();

      await flushPromises();

      expect(spyPush).not.toHaveBeenCalled();
    });
  });
});
