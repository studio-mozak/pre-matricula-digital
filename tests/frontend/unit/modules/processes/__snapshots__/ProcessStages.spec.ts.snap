// Vitest Snapshot v1

exports[`ProcessStages > should match snapshot 1`] = `
<main
  class="container"
  style="max-width: 740px;"
>
  <h2
    class="font-muli-20-primary"
  >
    Crie o processo de pré-matrícula
  </h2>
  <p>
     Crie os períodos de pré-matrícula que deseja para o seu processo. Cada processo poderá conter mais de um período de pré-matrícula, sendo as opções: um período de rematrícula e/ou um de matrícula. Você também pode escolher as escolas que aparecerão disponíveis por distância. 
  </p>
  <!--v-if-->
  <div>
    <form
      autocomplete="off"
      class="row"
    >
      
      
      
      
      <div
        class="col-12 mt-5"
      >
        <!--v-if-->
        <div
          class="row"
        >
          <div
            class="col-12"
          >
            <div
              class="d-flex"
            >
              <x-field-stub
                allowclear="false"
                bind="[object Object]"
                containerclass="w-100 mr-4"
                disabled="false"
                errors="false"
                label="Tipo de período"
                loading="false"
                modelvalue="REGISTRATION"
                name="stages[0].type"
                options="[object Object],[object Object],[object Object]"
                rules="required"
                searchable="false"
                type="SELECT"
              />
              <div
                class="d-flex align-items-center mt-3"
              >
                <button
                  class="x-btn x-btn-item non-selectable no-outline x-btn--standard x-btn--rectangle x-btn--actionable pmd-focusable pmd-hoverable x-btn--no-uppercase border-danger text-danger"
                  role="button"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="x-btn__content text-center d-flex justify-content-center no-wrap text-no-wrap"
                  >
                    <span
                      class="block"
                    >
                      Remover
                    </span>
                  </span>
                  <transition-stub
                    appear="false"
                    css="true"
                    name="x-transition--fade"
                    persisted="false"
                  >
                    <!---->
                  </transition-stub>
                </button>
              </div>
            </div>
          </div>
          <x-field-stub
            allowclear="false"
            bind="[object Object]"
            containerclass="col-12"
            disabled="false"
            errors="false"
            label="Opções de escola por raio <span style='font-size: 14px'>(?)</span>"
            loading="false"
            modelvalue="1500"
            name="stages[0].radius"
            options=""
            rules="numeric"
            searchable="false"
            type="NUMBER"
          />
          <div
            class="mb-2"
          >
            <x-field-stub
              allowclear="false"
              bind="[object Object]"
              containerclass="col-12 form-group"
              disabled="false"
              errors="false"
              label="Aceitar inscrições"
              loading="false"
              modelvalue="NONE"
              name="stages[0].restrictionType"
              options="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
              searchable="false"
              sublabel="Permite apenas inscrições de alunos(as) que possuam uma matrícula ativa no ano anterior"
              type="RADIO"
            />
          </div>
          <x-field-stub
            allowclear="false"
            bind="[object Object]"
            containerclass="col-12 form-group"
            disabled="false"
            errors="false"
            label="Permitir selecionar uma escola para lista de espera"
            loading="false"
            modelvalue="true"
            name="stages[0].allowWaitingList"
            options=""
            searchable="false"
            sublabel="Possibilita que no momento da pré-matrícula seja selecionada uma escola para lista de espera"
            type="CHECKBOX"
          />
          <x-field-stub
            allowclear="false"
            bind="[object Object]"
            containerclass="col-12 form-group"
            disabled="false"
            errors="false"
            label="Permitir se inscrever somente na escola atual do aluno"
            loading="false"
            modelvalue="false"
            name="stages[0].renewalAtSameSchool"
            options=""
            searchable="false"
            sublabel="Só possibilitará a inscrição caso exista vaga na escola atual do aluno"
            type="CHECKBOX"
          />
          <x-field-stub
            allowclear="false"
            bind="[object Object]"
            containerclass="col-12 form-group"
            disabled="false"
            errors="false"
            label="Permitir buscar por cadastros do i-Educar"
            loading="false"
            modelvalue="false"
            name="stages[0].allowSearch"
            options=""
            searchable="false"
            sublabel="Possibilitará buscar pelo aluno no i-Educar para reaproveitar os dados"
            type="CHECKBOX"
          />
          <x-field-stub
            allowclear="false"
            bind="[object Object]"
            containerclass="col-12 col-md-6"
            disabled="false"
            errors="false"
            label="Data Inicial"
            loading="false"
            modelvalue="2021-02-17 14:00:00"
            name="stages[0].startAt"
            options=""
            rules="required|date"
            searchable="false"
            type="DATE"
          />
          <x-field-stub
            allowclear="false"
            bind="[object Object]"
            containerclass="col-12 col-md-6"
            disabled="false"
            errors="false"
            label="Hora Inicial"
            loading="false"
            modelvalue="2021-02-17 14:00:00"
            name="stages[0].startHourAt"
            options=""
            rules="required|hour"
            searchable="false"
            type="TIME"
          />
          <x-field-stub
            allowclear="false"
            bind="[object Object]"
            containerclass="col-12 col-md-6"
            disabled="false"
            errors="false"
            label="Data Final"
            loading="false"
            modelvalue="2021-03-07 23:59:00"
            name="stages[0].endAt"
            options=""
            rules="required|date"
            searchable="false"
            type="DATE"
          />
          <x-field-stub
            allowclear="false"
            bind="[object Object]"
            containerclass="col-12 col-md-6"
            disabled="false"
            errors="false"
            label="Hora Final"
            loading="false"
            modelvalue="2021-03-07 23:59:00"
            name="stages[0].endHourAt"
            options=""
            rules="required|hour"
            searchable="false"
            type="TIME"
          />
          <x-field-stub
            allowclear="false"
            bind="[object Object]"
            containerclass="col-12"
            disabled="false"
            errors="false"
            label="Observações e lista de documentos a serem entregues na matrícula"
            loading="false"
            modelvalue=""
            name="stages[0].observation"
            options=""
            searchable="false"
            type="RICH_TEXT"
          />
        </div>
        <div
          class="row"
          style="display: none;"
        >
          <div
            class="col-12"
          >
            <p>
               O tipo de período 'Rematrícula' irá validar a compatibilidade de rematrícula, ou seja, validações de compatibilidade com os dados existentes no i-Educar permitindo a rematrícula somente de alunos que já estudam na rede. 
            </p>
          </div>
        </div>
      </div>
      
      <div
        class="col-12 col-md-6 offset-md-3 mt-5"
      >
        <div
          class="d-flex justify-content-center"
        >
          <button
            class="x-btn x-btn-item non-selectable no-outline x-btn--standard x-btn--rectangle bg-primary text-white x-btn--actionable pmd-focusable pmd-hoverable x-btn--no-uppercase w-100"
            data-test="add-stage-btn"
            role="button"
            tabindex="0"
            type="button"
          >
            <span
              class="x-btn__content text-center d-flex justify-content-center no-wrap text-no-wrap"
            >
              <span
                class="block"
              >
                Adicionar outro período
              </span>
            </span>
            <transition-stub
              appear="false"
              css="true"
              name="x-transition--fade"
              persisted="false"
            >
              <!---->
            </transition-stub>
          </button>
        </div>
      </div>
      
      <div
        class="col-12"
      >
        <div
          class="row mt-5"
        >
          
          <div
            class="offset-md-3 col-md-3"
          >
            <button
              class="x-btn x-btn-item non-selectable no-outline x-btn--outline x-btn--rectangle text-gray-500 border-gray-500 x-btn--actionable pmd-focusable pmd-hoverable x-btn--no-uppercase w-100"
              data-test="button-back"
              role="button"
              style="font-size: 20px;"
              tabindex="0"
              type="button"
            >
              <span
                class="x-btn__content text-center d-flex justify-content-center no-wrap text-no-wrap"
              >
                <span
                  class="block"
                >
                  Voltar
                </span>
              </span>
              <transition-stub
                appear="false"
                css="true"
                name="x-transition--fade"
                persisted="false"
              >
                <!---->
              </transition-stub>
            </button>
          </div>
          <div
            class="col-md-3 mt-3 mt-md-0"
          >
            <button
              class="x-btn x-btn-item non-selectable no-outline x-btn--standard x-btn--rectangle bg-primary text-white x-btn--actionable pmd-focusable pmd-hoverable x-btn--no-uppercase btn btn-block btn-primary w-100 h-100 flex-row justify-content-center"
              disabled="false"
              role="button"
              style="font-size: 20px;"
              tabindex="0"
              type="button"
            >
              <span
                class="x-btn__content text-center d-flex justify-content-center no-wrap text-no-wrap"
              >
                <span
                  class="block"
                >
                  Prosseguir
                </span>
              </span>
              <transition-stub
                appear="false"
                css="true"
                name="x-transition--fade"
                persisted="false"
              >
                <!---->
              </transition-stub>
            </button>
          </div>
          
        </div>
      </div>
    </form>
  </div>
</main>
`;
